src/app/inventory/dimCsvService.factory.ts

```diff
@@ -55,194 +59,230 @@ function buildSocketString(sockets: DimSockets): string {
       )
   );
 
-  return _.flatten(socketItems).join(',');
+  return _.flatten(socketItems);
 }
 
-function buildNodeString(nodes) {
-  let data = '';
-  nodes.forEach((node) => {
-    if (FILTER_NODE_NAMES.includes(node.name)) {
-      return;
-    }
-    data += node.name;
-    if (node.activated) {
-      data += '*';
-    }
-    data += ',';
-  });
-
-  return data;
+function buildNodeString(nodes: DimGridNode[]): string[] {
+  return _.compact(
+    nodes.map((node) => {
+      if (FILTER_NODE_NAMES.includes(node.name)) {
+        return;
+      }
+      return node.activated ? `${node.name}*` : node.name;
+    })
+  );
 }
 
-function downloadGhost(items, nameMap) {
-  const header = 'Name,Hash,Id,Tag,Tier,Owner,Locked,Equipped,Perks\n';
-
-  let data = '';
-  items.forEach((item) => {
-    data += `"${item.name}",`;
-    data += `"${item.hash}",`;
-    data += `${item.id},`;
-    data += `${item.dimInfo.tag || ''},`;
-    data += `${item.tier},`;
-    data += `${nameMap[item.owner]},`;
-    data += `${item.locked},`;
-    data += `${item.equipped},`;
-
-    if (item.talentGrid) {
-      data += buildNodeString(item.talentGrid.nodes);
-    } else if (item.sockets) {
-      data += buildSocketString(item.sockets);
-    }
+function downloadGhost(items: DimItem[], nameMap: { [key: string]: string }) {
+  // We need to always emit enough columns for all perks
+  const maxPerks =
+    _.max(
```

Comment:
Calculating the max # of columns for a collection of items should be extracted to a function.